src/Microsoft.ML.Transforms/Text/NgramHashingTransformer.cs

```diff
@@ -288,506 +267,456 @@ public ColInfoEx(ModelLoadContext ctx)
 
                 // *** Binary format ***
+                // size of Inputs
+                // string[] Inputs;
+                // string Output;
                 // int: NgramLength
                 // int: SkipLength
                 // int: HashBits
                 // uint: Seed
                 // byte: Rehash
                 // byte: Ordered
                 // byte: AllLengths
+                Contracts.Assert(Inputs.Length > 0);
+                ctx.Writer.Write(Inputs.Length);
+                for (int i = 0; i < Inputs.Length; i++)
+                    ctx.SaveNonEmptyString(Inputs[i]);
+                ctx.SaveNonEmptyString(Output);
 
-                Contracts.Assert(NgramLength > 0);
+                Contracts.Assert(0 < NgramLength && NgramLength <= NgramBufferBuilder.MaxSkipNgramLength);
                 ctx.Writer.Write(NgramLength);
-                Contracts.Assert(SkipLength >= 0);
+                Contracts.Assert(0 <= SkipLength && SkipLength <= NgramBufferBuilder.MaxSkipNgramLength);
+                Contracts.Assert(NgramLength + SkipLength <= NgramBufferBuilder.MaxSkipNgramLength);
                 ctx.Writer.Write(SkipLength);
                 Contracts.Assert(1 <= HashBits && HashBits <= 30);
                 ctx.Writer.Write(HashBits);
                 ctx.Writer.Write(Seed);
-                ctx.Writer.WriteBoolByte(Rehash);
+                ctx.Writer.WriteBoolByte(RehashUnigrams);
                 ctx.Writer.WriteBoolByte(Ordered);
                 ctx.Writer.WriteBoolByte(AllLengths);
             }
         }
 
-        internal const string Summary = "Produces a bag of counts of ngrams (sequences of consecutive values of length 1-n) in a given vector of keys. "
-            + "It does so by hashing each ngram and using the hash value as the index in the bag.";
-
-        public const string LoaderSignature = "NgramHashTransform";
-        private static VersionInfo GetVersionInfo()
-        {
-            return new VersionInfo(
-                modelSignature: "HASHGRAM",
-                // verWrittenCur: 0x00010001, // Initial
-                verWrittenCur: 0x00010002, // Invert hash key values, hash fix
-                verReadableCur: 0x00010002,
-                verWeCanReadBack: 0x00010002,
-                loaderSignature: LoaderSignature,
-                loaderAssemblyName: typeof(NgramHashingTransformer).Assembly.FullName);
-        }
-
-        private readonly Bindings _bindings;
-        private readonly ColInfoEx[] _exes;
-
+        private readonly ImmutableArray<ColumnInfo> _columns;
         private readonly VBuffer<ReadOnlyMemory<char>>[] _slotNames;
         private readonly ColumnType[] _slotNamesTypes;
 
-        private const string RegistrationName = "NgramHash";
+        public NgramHashingTransformer(IHostEnvironment env, params ColumnInfo[] columns) :
+            base(Contracts.CheckRef(env, nameof(env)).Register(nameof(NgramHashingTransformer)))
```

Comment:
From the constructor below it seems that this is a trainable transformation, why does it have a public constructor then?

Reply:
Still don't understand why you have a public constructor for a trainable transform?

---
In reply to: [238786341](https://github.com/dotnet/machinelearning/pull/1811#discussion_r238786341) [](ancestors = 238786341)